src/attribute.ts

```diff
@@ -1,80 +1,193 @@
-// In the future, this class will be used for
-// transforms, default values, etc.
+import { JSORMBase } from './model'
 
-import Model from './model';
-import Config from './configuration';
+export type Attr<T> = { (): T } | { new (...args: any[]): T & object }
 
-export default class Attribute {
-  name: string;
+export type AttrType<T> = Attr<T>
 
-  persist: boolean = true;
-  isAttr: boolean = true;
-  isRelationship: boolean = false;
+export interface AttrRecord<T> {
+  name?: string | symbol
+  type?: AttrType<T>
+  persist? : boolean
+}
 
-  constructor(opts?: attributeOptions) {
-    if (opts && opts.hasOwnProperty('persist')) {
-      this.persist = opts.persist;
-    }
+export function attr<T=any>(options? : AttrRecord<T>) : Attribute<T> {
```

Comment:
I'm open to this being an experience thing, but it's really hard for me to grok what's going on when the type definitions and functions are spliced together. Could we put all types at the top, all functions/classes at the bottom?

Reply:
Yeah going to clean all this up at some point before release.